crates/protocol/src/version.rs

```diff
@@ -0,0 +1,45 @@
+const SNAPSHOT_FLAG: i32 = 1 << 30;
+
+/// A protocol version that can be encoded in the numeric format.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum ProtocolVersion {
```

Comment:
I am not sure if this enum is sound: you can manualy create it as `SnapshotFlag::Stable(1 << 30 + 1)` and break the invariant.

I suggest turning this into a newtype-`i32` with methods like `version()` and `is_snapshot()` checking the stored value.

Alternatively you could crate separate newtype structs like:

```rust
pub struct Stable(i32);
pub struct Snapshot(i32);
```

and using them as

```rust
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ProtocolVersion {
	
Attach files by dragging & dropping, selecting or pasting them.
    Stable(Stable),
    Snapshot(Snapshot
}
```

But it would be more verbose for no useful purpose.

Reply:
Let's not get too pedantic

Reply:
It's still quite counter-intuitive and not Rust-idiomatic.

I suggest just a simple struct:

```
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct ProtocolVersion(pub i32);

impl ProtocolVersion {
    fn version() -> i32 { ... } // should convert on snapshot
    fn is_snapshot() -> bool { ... }
}
```

It both has a better memory layout and cannot be misused.

Reply:
Done